runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -346,11 +346,20 @@ public V get(@ConflictParameter Object key) {
     <I extends Comparable<I>>
     Collection<Entry<K, V>> getByIndex(@Nonnull IndexName indexName, I indexKey) {
         String secondaryIndex = indexName.get();
-        Map<K, V> res = secondaryIndexes.get(secondaryIndex).get(indexKey);
+        if (secondaryIndexes.containsKey(secondaryIndex)) {
+            Map<Comparable, Map<K, V>> secondaryMap = secondaryIndexes.get(secondaryIndex);
+            if (secondaryMap != null) {
+                Map<K, V> res = secondaryMap.get(indexKey);
 
-        return res == null ?
-                Collections.emptySet():
-                new HashSet<>(res.entrySet());
+                return res == null ?
+                        Collections.emptySet() :
+                        new HashSet<>(res.entrySet());
+            }
+        }
+
+        // If index is not specified, the lookup by index API must fail.
```

Comment:
A suggestion: what about doing the secondaryMap check on top of the method (as it is really the pre condition) and throwing the exceptions right away if it doesn't exist. This eliminates the need for other if condition at L349 and is more straightforward.